### PR TITLE
feat(plugins): runtime install — server + frontend + preset (Phase A+B+C+preset / #1043 C-2)

### PR DESCRIPTION
**Pull request by isamu**
## Summary

- Server-side foundation for "install npm plugin → restart → it works" (#1043 C-2). After this PR, plugin tarballs in \`~/mulmoclaude/plugins/\` are extracted on boot, their MCP \`TOOL_DEFINITION\` loaded, and exposed to the MCP child process. Curl-testable end-to-end.
- Frontend Vue View dynamic load lands in the next PR (Phase C).
- Architecture and 5-phase rollout: \`plans/feat-plugin-c2-impl.md\` (committed in this branch). Technical feasibility was retired in 5 spikes documented at https://github.com/receptron/mulmoclaude/issues/1043#issuecomment-4357950541.

## Items to Confirm / Review

- **Plugin merge into MCP**: \`server/agent/mcp-server.ts\` now awaits a module-level \`runtimeReady\` Promise before serving \`tools/list\` and \`tools/call\`. Top-level await would be cleaner but the Docker build path uses tsx with cjs output, which fails to parse TLA at module init. The Docker smoke test passes with the new shape — early requests queue on the promise.
- **Auto-include in active set**: runtime plugins are added to the MCP active tool set regardless of \`role.availablePlugins\`. Rationale: the user explicitly installed the plugin, so every role can use it. Static plugins still go through the role gate via \`PLUGIN_NAMES\` env (set by the parent based on \`getActivePlugins(role)\`). Confirm this is the right call vs. requiring per-role opt-in.
- **Collision policy**: static names always win; runtime-vs-runtime first-loaded wins. Skipped collisions log a WARN and surface in \`RegisterResult.collisions\`. The web UI (Phase D) will surface this to the user during install.
- **Dispatch body shape**: \`POST /api/plugins/runtime/:pkg/dispatch\` accepts the args directly as the body (no envelope) — same convention as static plugin routes (\`server/api/routes/plugins.ts\`), so \`postJson(endpoint, args)\` in mcp-server works unchanged.
- **Asset route security**: \`/api/plugins/runtime/:pkg/:version/*\` normalises the path and rejects anything outside the plugin's cache directory. Confirm the path-traversal check is correct.
- **Top-level-await avoidance**: mcp-server.ts uses \`runtimeReady\` Promise plus then-chains rather than \`await\`. The result is: tools/list and tools/call wait for runtime plugin loading; ping and initialize don't. Reviewers should confirm the JSON-RPC ordering is correct (initialize → tools/list flow).

## User Prompt

> いま、mulmo claudeの機能追加のpluginを実装しているんだよね。ブリッジ関連は、それとは独立しているから関係ないよね？
>
> はい、技術的にできるかどうかが重要。そのうえでユーザが明示的にインストールしたものは動くべきなので、permission modelは不要では？ web上で追加したものが、動的に、もしくはreloadしたときに動けば良い。ただし、これはnpxで起動時にも動くべきなのでworkspaceに展開されるのかな。tgz状態でworkspaceにおいて、起動時に展開（キャッシュあり）とか？
>
> はい、まずissueにこれらを記載したうえでlocalでテストをお願いします
>
> 全部検証して
>
> A → 本実装計画を書いて → ブランチで進めて、PR は壊れなければ細かくてもよいけど、まとめないと動かないならまとめて

### Workspace layout

\`\`\`
~/mulmoclaude/
  plugins/
    plugins.json                     ← install ledger (LedgerEntry[])
    @gui-chat-plugin-weather-0.1.0.tgz
    @mulmochat-plugin-quiz-0.4.0.tgz
    .cache/                          ← extracted on first boot
      @gui-chat-plugin/weather/0.1.0/
        package.json
        dist/{index.js, vue.js, ...}
\`\`\`

### Server boot flow

1. \`readLedger()\` → \`LedgerEntry[]\`
2. Extract any tgz missing from cache (cache hit = skip; \`tar -xzf --strip-components=1\` via \`execFileSync\` to bypass shell parsing)
3. Dynamic-import \`dist/index.js\` (resolved via \`exports["."].import\` → \`module\` → \`main\`) and pull \`TOOL_DEFINITION\`
4. \`registerRuntimePlugins(MCP_PLUGIN_NAMES, runtimePlugins)\` applies the collision policy (static wins; runtime first-loaded wins)
5. \`mcp-server.ts\` merges runtime tools into \`ALL_TOOLS\`, auto-adds them to the active set, and exposes via \`tools/list\` / \`tools/call\`

### What's new

- \`server/utils/files/plugins-io.ts\` — ledger I/O via \`writeJsonAtomic\`
- \`server/plugins/runtime-loader.ts\` — extract + import flow
- \`server/plugins/runtime-registry.ts\` — in-process registry + collision policy
- \`server/api/routes/runtime-plugin.ts\` — list / dispatch / asset endpoints

### What's modified

- \`server/agent/mcp-server.ts\` — await \`runtimeReady\` before tools/list and tools/call
- \`server/index.ts\` — \`loadRuntimePlugins()\` + register at boot, mount route
- \`server/workspace/paths.ts\` — \`plugins\` / \`pluginCache\` / \`pluginsLedger\`
- \`src/config/workspacePaths.ts\` — \`pluginsLedger\`
- \`src/config/apiRoutes.ts\` — \`plugins.runtimeList\` / \`runtimeDispatch\` / \`runtimeAsset\`

## Test plan

- [x] \`yarn typecheck\` clean
- [x] \`yarn lint\` clean (3 pre-existing unrelated warnings)
- [x] \`yarn test\` passes including the Docker smoke test that exercises tsx-cjs
- [x] \`yarn build\` clean
- [x] 16 new unit tests (loader fixtures + registry collision policy)
- [x] Manual e2e: \`@gui-chat-plugin/browse\` + \`@gui-chat-plugin/weather\` in actual workspace → \`loadRuntimePlugins()\` → \`registerRuntimePlugins()\` → both registered, no collisions, \`MCP_PLUGIN_NAMES\` exclusion works
- [ ] CI green on this PR
- [ ] Manual smoke: full server boot with installed plugins (deferred to next PR — frontend doesn't render runtime View yet)

## Out of scope (future PRs)

- **Phase C**: frontend dynamic load — importmap, Vue bridge, runtime view registration
- **Phase D**: install CLI (\`yarn plugin:install <pkg>\`) + web UI install button
- **Phase E**: production build path (importmap target hashing in vite build), docker mount of cache dir, e2e tests

## Plan / tracking

- Plan: \`plans/feat-plugin-c2-impl.md\`
- Tracking: #1043 (umbrella for plugin SDK / dynamic install / marketplace)
- Phase 1a (#1064, merged): manifest-driven plugin registration
- Phase 1b (#1067, closed): bridge protocol semver — out of scope for #1043

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Install, list, and run runtime npm-packaged plugins with on-disk persistence.
  * Frontend support for dynamically loading plugin UI and styles so plugins appear in-app.
  * New API endpoints to list plugins, dispatch plugin calls, and serve plugin assets.

* **Documentation**
  * Added end-to-end guide for creating, installing, and validating runtime plugins.

* **Tests**
  * Added unit, integration, and end-to-end tests covering loader, registry, asset serving, and security edge cases.

### CODE REVIEW

File: server/agent/mcp-server.ts [hunk 1]
```diff
@@ -91,12 +93,35 @@ const mcpToolDefs: Record<string, ToolDef> = Object.fromEntries(
...
   ...Object.fromEntries(PLUGIN_DEFS.map((def) => [def.name, fromPackage(def, TOOL_ENDPOINTS[def.name])])),
 };
 
-const tools = PLUGIN_NAMES.map((name) => ALL_TOOLS[name]).filter(Boolean);
+let activeNames: string[] = [...PLUGIN_NAMES];
+let tools: ToolDef[] = activeNames.map((name) => ALL_TOOLS[name]).filter(Boolean);
+
+const runtimeReady: Promise<void> = (async () => {
+  const runtime = await loadRuntimePlugins();
+  registerRuntimePlugins(MCP_PLUGIN_NAMES, runtime);
```
Comment (isamu) [inline server/agent/mcp-server.ts]: `registerRuntimePlugins()` is only given `MCP_PLUGIN_NAMES`, but `ALL_TOOLS` also contains the MCP tool defs from `mcpToolDefs` (`readXPost`, `searchX`, `notify`). A runtime plugin using one of those names will still overwrite the built-in entry at `ALL_TOOLS[plugin.definition.name] = ...` below, so the stated "static built-ins win every collision" policy is not actually enforced here. Please include the MCP tool names in the static collision set or guard the assignment as well.

File: server/agent/mcp-server.ts [hunk 2]
```diff
@@ -400,17 +425,23 @@ process.stdin.on("data", (chunk: Buffer) => {
...
-            description: toolDef.description,
-            inputSchema: toolDef.inputSchema,
-          })),
-        },
-      });
+      // Await runtime-plugin load before responding so workspace-
+      // installed tools appear in the very first list call. Without
+      // this, an early tools/list could miss them and the LLM would
+      // never call them this session.
+      runtimeReady.then(() =>
```
Comment (isamu) [inline server/agent/mcp-server.ts:457]: This awaits `runtimeReady` without any rejection path. `loadRuntimePlugins()` can still reject on filesystem errors that `loadJsonFile()` rethrows (e.g. unreadable ledger / EACCES), and in that case `tools/list` never responds, so the MCP child looks hung during handshake. Please add a `.catch(...)` here or make `runtimeReady` swallow boot-time loader failures before the request path awaits it.

File: server/plugins/runtime-loader.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+async function loadOne(entry: LedgerEntry): Promise<RuntimePlugin | null> {
+  const tgzAbs = path.join(WORKSPACE_PATHS.plugins, entry.tgz);
+  if (!existsSync(tgzAbs)) {
+    log.warn(LOG_PREFIX, "tgz missing — skipping", { name: entry.name, tgz: entry.tgz });
+    return null;
+  }
+  const cachePath = path.join(WORKSPACE_PATHS.pluginCache, entry.name, entry.version);
+  if (!existsSync(cachePath)) {
```
Comment (isamu) [inline server/plugins/runtime-loader.ts]: Using `existsSync(cachePath)` as the only cache-validity check makes partial extractions sticky. If `tar` is interrupted or writes an incomplete tree, the directory already exists, so future boots will skip re-extraction forever and the plugin stays broken until manual cleanup. Consider a completion marker / checksum, or re-extract when the import step fails.

**Review by isamu — COMMENTED** [review]
Found three blocking issues in the runtime plugin foundation.

**Review by isamu — COMMENTED** [review]
CODEX VERDICT: CHANGES REQUESTED
- Runtime plugins can still shadow built-in MCP tools because the collision filter only covers `MCP_PLUGIN_NAMES`, not the full static tool set.
- `tools/list` waits on `runtimeReady` without a rejection path, so a boot-time filesystem error can hang the MCP child.
- The tgz cache is treated as valid just because the directory exists, so a partial extraction can become a permanent broken cache until manual cleanup.

**Comment by isamu** [comment]
CODEX VERDICT: CHANGES REQUESTED

- `server/workspace/memory/topic-migrate.ts:53` returns before `resetStaging()` when there are no entries. That leaves any pre-existing `conversations/memory.next` in place, so a stale staging tree from an earlier failed run can survive and be swapped later.
- `server/workspace/memory/topic-migrate.ts:155` writes `writtenSlug` directly into `path.join()` without validating it. A malformed `MemoryClusterer` output can escape `memory.next` (`../...`) and write outside the staging directory.

**Comment by isamu** [comment]
Re: codex iter-2 second verdict (re `server/workspace/memory/topic-migrate.ts`):

Those two findings target a file that is **not part of this PR**. `topic-migrate.ts` landed on `main` via PR #1072 (the memory-topic-restructure branch that merged just before iteration 2 started), and was pulled in via the routine `git merge origin/main` step before the iter-1 push. The diff for this PR's branch (`gh pr diff 1077`) does not touch it.

Filing them here is a misclassification — they belong on a separate issue / follow-up PR against `main`. Skipping in this PR's review loop. The valid in-scope finding from iter-2 (asset route bearer-auth exemption) is fixed in the next commit.

File: server/index.ts [hunk 4]
```diff
@@ -126,12 +130,26 @@ app.use(requireSameOrigin);
...
+// for the same reason: the frontend dynamic-imports plugin assets
+// (`import("/api/plugins/runtime/<pkg>/<ver>/dist/vue.js")`) and the
+// browser cannot attach Authorization headers to those module
+// requests. The pattern "4+ segments past /plugins/runtime/" only
+// matches asset GETs — `/plugins/runtime/list` (3 segments) and
+// `/plugins/runtime/<pkg>/dispatch` (3 segments) still require auth.
+// Path traversal is hardened separately by `resolveWithinRoot` in
+// the asset route handler.
 // The CSRF origin check + loopback-only binding still apply.
+const RUNTIME_PLUGIN_ASSET_PATH_RE = /^\/plugins\/runtime\/[^/]+\/[^/]+\//;
```
Comment (isamu) [inline server/index.ts:167]: This exemption is broader than the browser asset case: `runtime-plugin.ts` decodes `req.params.pkg` / `req.params.version` before `path.join(...)`. A request with percent-encoded `../` in either segment can still pass this middleware check, and `resolveWithinRoot()` only constrains the wildcard `subPath`, not the joined root. That makes the new GET-only bypass an unauthenticated path to the static mount.

**Review by isamu — COMMENTED** [review]
One regression to call out: the GET exemption is correct for list/dispatch separation, but it also bypasses auth for crafted asset URLs that can smuggle traversal through decoded `pkg` / `version` parameters.

**Comment by isamu** [comment]
CODEX VERDICT: CHANGES REQUESTED
- The `GET` exemption in `server/index.ts` does correctly keep `/api/plugins/runtime/list` and `/api/plugins/runtime/:pkg/dispatch` behind bearer auth.
- The new regex also lets unauthenticated GETs reach the runtime asset handler with crafted `:pkg` / `:version` values; those params are decoded later, so percent-encoded traversal can still escape the intended plugin cache root because `resolveWithinRoot()` only constrains the wildcard subpath.

**Comment by isamu** [comment]
CODEX VERDICT: LGTM

Reviewed the local diff for `server/api/routes/runtime-plugin.ts`, `src/config/apiRoutes.ts`, and `test/api/routes/test_runtimePluginRoot.ts`.

Verified:
- `resolvePluginRoot(pkg, version)` realpath-anchors the candidate under `WORKSPACE_PATHS.pluginCache`, so decoded `../` traversal in either segment is rejected before `resolveWithinRoot()` runs.
- The runtime asset route syntax `/{*splat}` is valid under the repo’s Express 5 router dependency (`router@2.2.0` / `path-to-regexp@8.4.0`), and `req.params.splat` is populated for multi-segment asset paths.

No blocking concerns.

**Comment by isamu** [comment]
## 動作確認手順

このPR は server-side 完結（frontend 描画は Phase C）なので、確認は MCP 経由でツールが見える + route が curl で叩ける + セキュリティガードが効く の3点。

### 前提

```bash
ls ~/mulmoclaude/plugins/
cat ~/mulmoclaude/plugins/plugins.json
```

ファイルが無ければ手動で1個入れる（install CLI は Phase D なので暫定）:

```bash
mkdir -p ~/mulmoclaude/plugins
cd "$(mktemp -d)" && npm pack @gui-chat-plugin/weather && \
  mv gui-chat-plugin-weather-*.tgz ~/mulmoclaude/plugins/
cat > ~/mulmoclaude/plugins/plugins.json <<JSON
[{"name":"@gui-chat-plugin/weather","version":"0.1.0","tgz":"gui-chat-plugin-weather-0.1.0.tgz","installedAt":"$(date -u +%Y-%m-%dT%H:%M:%SZ)"}]
JSON
```

### 1. サーバ起動 → boot ログで loader が走るか

```bash
git switch feat/plugin-c2-server
yarn dev 2>&1 | grep -i "plugins/runtime"
```

期待ログ:

```
INFO [plugins/runtime] extracted name=@gui-chat-plugin/weather version=0.1.0
INFO [plugins/runtime] loaded requested=1 succeeded=1
INFO [plugins/runtime] registered runtime plugins registered=1 collisions=0
```

完了マーカー確認:

```bash
ls ~/mulmoclaude/plugins/.cache/@gui-chat-plugin/weather/0.1.0/.extract-complete
```

### 2. List endpoint（要 auth）

```bash
TOKEN=$(cat ~/mulmoclaude/.session-token)
curl -s -H "Authorization: Bearer $TOKEN" \
     -H "Origin: http://localhost:5173" \
     http://localhost:3001/api/plugins/runtime/list | jq
```

期待:

```json
{
  "plugins": [{
    "name": "@gui-chat-plugin/weather",
    "version": "0.1.0",
    "toolName": "fetchWeather",
    "description": "Fetch weather forecast information ...",
    "assetBase": "/api/plugins/runtime/%40gui-chat-plugin%2Fweather/0.1.0"
  }]
}
```

### 3. Asset endpoint（auth 不要 — browser dynamic import 用）

```bash
curl -s -H "Origin: http://localhost:5173" \
     "http://localhost:3001/api/plugins/runtime/%40gui-chat-plugin%2Fweather/0.1.0/dist/index.js" \
     | head -3
```

期待: `import { ... } from "./samples-...js";` のような JS ヘッダ

### 4. Dispatch endpoint（要 auth、execute 呼び出し）

```bash
curl -s -H "Authorization: Bearer $TOKEN" \
     -H "Origin: http://localhost:5173" \
     -H "Content-Type: application/json" \
     -X POST \
     -d '{"areaCode":"130000"}' \
     "http://localhost:3001/api/plugins/runtime/%40gui-chat-plugin%2Fweather/dispatch" | jq | head -20
```

期待: 気象庁API から東京の天気が返ってくる

### 5. セキュリティ: encoded traversal が 404 になる

```bash
curl -sI -H "Origin: http://localhost:5173" \
     "http://localhost:3001/api/plugins/runtime/%2E%2E%2F%2E%2E%2Fetc/passwd/dist/index.js" \
     | head -1
```

期待: `HTTP/1.1 404 Not Found` （`resolvePluginRoot` が cache root anchor で reject）

### 6. MCP 経由で LLM が tool を見えてる確認

`yarn dev` で起動したまま GUI で session を開き、「東京の天気おしえて」と入力。`fetchWeather` tool 呼び出しがサーバ側ログに出れば OK:

```bash
yarn dev 2>&1 | grep -i fetchWeather
```

または agent の MCP 通信の `tools/list` レスポンスを直接見たいなら:

```bash
yarn dev:debug 2>&1 | grep -A2 '"name":"fetchWeather"'
```

### 最低限の動作確認

**1 → 2 → 3 → 5** で一通り動作 + セキュリティが確認できる。全部やるなら 4 と 6 も。

**Comment by isamu** [comment]
## 動作確認手順 (Phase C 込みで更新)

Phase C を branch に追加 commit (`f84acac7`) しました。**実際のブラウザで runtime plugin の View が描画される end-to-end フロー** が確認できます。

### 前提

```bash
ls ~/mulmoclaude/plugins/
cat ~/mulmoclaude/plugins/plugins.json
```

無ければ手動で1個入れる:

```bash
mkdir -p ~/mulmoclaude/plugins
cd "$(mktemp -d)" && npm pack @gui-chat-plugin/weather && \
  mv gui-chat-plugin-weather-*.tgz ~/mulmoclaude/plugins/
cat > ~/mulmoclaude/plugins/plugins.json <<JSON
[{"name":"@gui-chat-plugin/weather","version":"0.1.0","tgz":"gui-chat-plugin-weather-0.1.0.tgz","installedAt":"$(date -u +%Y-%m-%dT%H:%M:%SZ)"}]
JSON
```

### 1. サーバ起動 → boot ログ

```bash
git switch feat/plugin-c2-server && yarn dev 2>&1 | grep "plugins/runtime"
```

期待:
```
[plugins/runtime] loaded requested=1 succeeded=1
[plugins/runtime] registered runtime plugins registered=1 collisions=0
```

### 2. ブラウザで実際に開く

[http://localhost:5173](http://localhost:5173) を開いて、devtools → Network で:
- `/api/plugins/runtime/list` (200, plugins 配列が返る)
- `/api/plugins/runtime/<pkg>/<ver>/dist/vue.js` (200)
- `/api/plugins/runtime/<pkg>/<ver>/dist/style.css` (200)

devtools → Console で:
```js
const m = await import("vue");
console.log(m.version);  // 3.5.33
```

### 3. Chat で fetchWeather を実際に呼ぶ

GUI で role を一つ選んで "東京の天気おしえて" と入力。

期待:
- LLM が `fetchWeather` tool を呼ぶ
- canvas に weather plugin の View が描画される（⛅ アイコン + Tailwind スタイル付き）
- preview pane (left side) に previewComponent が描画される

### 4. Phase C 自動検証スクリプト（任意）

dev server が起動した状態で:
```bash
npx tsx scripts/verify-phase-c.mts
```

期待最後の行: `[phase-c] SUCCESS`

検証内容:
- list endpoint が plugins 配列を返す
- 各 plugin の vue.js + style.css 取得 200
- dynamic import で plugin module ロード成功
- `HostVue === PluginVue` 確認 (同一 instance / 3.5.33)
- viewComponent + previewComponent 両方 export 済み

### 5. セキュリティ: encoded traversal が 404 (server side のまま)

```bash
curl -s -o /dev/null -w "%{http_code}\n" -H "Origin: http://localhost:5173" \
  "http://localhost:3001/api/plugins/runtime/%2E%2E%2F%2E%2E%2Fetc/passwd/dist/index.js"
```

期待: `404` (`resolvePluginRoot` が cache root anchor で reject)

---

**最低限**: 1 → 2 → 3 で実際の user 体験まで確認できる。CI が通る + 4 が SUCCESS なら机上検証も OK。

**Comment by isamu** [comment]
Review result: one remaining issue.

`server/plugins/runtime-loader.ts:164-170` still probes `entry.tgz` with `existsSync()` before the new `ensureInsideBase()` guard runs. That means a malformed ledger entry can still force a disk lookup outside `WORKSPACE_PATHS.plugins` before it is rejected, which is at odds with the stated “before touching disk” hardening. I’d move the base check ahead of the tgz existence probe, or validate/sanitize `entry.tgz` separately.

Verified the build-only importmap rewrite: `yarn build:client` succeeds, emits `dist/client/assets/runtime-vue-B0yQ6H_F.js`, and the built `dist/client/index.html` rewrites both the comment and the importmap target to that hashed asset.

Verdict: not yet LGTM.
